vignettes_src/workers.Rmd

```diff
@@ -0,0 +1,287 @@
+  "  ok <- rrq::rrq_task_wait(ids)",
+  "  stopifnot(ok)",
+  "  result <- rrq::rrq_task_results(ids)",
+  "  rrq::rrq_task_delete(ids)",
+  "  sum(unlist(result))",
+  "}")
+writeLines(code, "code.R")
+```
+
+We'll submit a hipercow task that runs this small piece of code:
+
+```{r, echo = FALSE, results = "asis"}
+r_output(readLines("code.R"))
+```
+
+Hopefully this is fairly self-explanatory, if a bit pointless.  Note that we delete the rrq tasks after completing them, which prevents rrq tasks accumulating.
+
+This code needs to be available in the hipercow environment run by the task:
+
+```{r}
+hipercow_environment_create("default", sources = "code.R")
+```
+
+We also need some workers - let's submit 4 of them:
+
+```{r}
+info <- hipercow_rrq_workers_submit(4)
+```
+
+now submit a task that will call this function:
+
+```{r}
+id <- task_create_expr(
+  example(16),
+  parallel = hipercow_parallel(use_rrq = TRUE))
+id
+```
+
+Here, `id` is a *hipercow* task identifier.  This sends an additional task to the queue, so now we have five things running on the cluster (four workers and one task).  The task runs, picks up the controller, then uses it do distribute a (trivial) calculation over the four workers.
+
+```{r}
+task_wait(id)
+task_result(id)
+```
+
+You can see from the worker logs here the tasks being split between the workers:
+
+```{r}
+rrq_worker_log_tail(n = 32)
+```
+
+This example is trivial, but you could submit 10 workers each using a 32 core node, and then use a single core task to farm out a series of large simulations across your bank of computers.  Or create a 500 single core workers (so ~25% of the cluster) and smash through a huge number of simulations with minimal overhead.
+
+# Tricks and tips
+
+This section will expand as we document patterns that have been useful.
+
+## Controlling the worker environment
+
+The workers will use the `rrq` environment if it exists, failing that the `default` environment.  So if you need different packages and sources loaded on the workers on your normal tasks, you can do this by creating a different environment
+
+```{r}
+hipercow_environment_create("rrq", packages = "cowsay")
+```
+
+**TODO**: *work out how to refresh this environment; I think that's just a message to send*
+
+You can submit your workers with any resources and parallel control you want (see `vignettes("parallel")` for details); pass these as `resources` and `parallel` to `hipercow_rrq_workers_submit()`.
+
+
+# General considerations
+
+## Stopping redundant workers
+
+By default, workers will live for 10 minutes after finishing their last task.  This means that most of the time that you use workers you can largely forget about cleanup.  If you want be polite and give up these resources early (this would be important if you wanted to launch new workers, or if you were using a large fraction of the cluster), you can tell them to stop immediately after completion of the last job by setting their timeout to zero.
```

Comment:
```suggestion
By default, workers will live for 10 minutes after finishing their last task.  This means that most of the time that you use workers you can largely forget about cleanup.  If you want to be polite and give up these resources early (this would be important if you wanted to launch new workers, or if you were using a large fraction of the cluster), you can tell them to stop immediately after completion of the last job by setting their timeout to zero.
```